### PR TITLE
Add memoryUsage for principal cache

### PR DESCRIPTION
**Pull request by t4niwa**
# Description

Add memoryUsage for principal cache

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Non-code changes (update documentation, pipeline, etc.)

### Flags

- [ ] Breaks backward compatibility
- [ ] Requires a documentation update
- [ ] Has untestable code

## Related issue/PR

**Delete this section if there are no issues or pull requests that relate to this pull request.**
- Fixes #_issue_
- Closes #_PR_

---

## Checklist

- [ ] Followed the guidelines in the CONTRIBUTING document
- [ ] Added prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]` in the PR title if necessary
- [ ] Tested and linted the code
- [ ] Commented the code
- [ ] Made corresponding changes to the documentation
- [ ] Passed all pipeline checking

## Checklist for maintainer

- Use `Squash and merge`
- Double-confirm the merge message has prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]`
- Delete the branch after merge

### CODE REVIEW

File: authorizerd.go [hunk 1]
```diff
@@ -467,9 +480,59 @@ func (a *authority) authorize(ctx context.Context, m mode, tok, act, res, query
 		return fmt.Sprintf("set token result. masked tok: %s, masked key: %s, act: %s, res: %s", maskToken(m, tok), maskCacheKey(key.String(), tok), act, res)
 	})
 	a.cache.SetWithExpire(key.String(), p, a.cacheExp)
+
+	principalCacheSize := principalCacheMemoryUsage(p) + (int64(len(key.String())) * 2)
```
[the remark this answers is not in the document]
Comment (fum1h1to) [reply authorizerd.go:487]: fix: [6d927ac](https://github.com/AthenZ/athenz-authorizer/pull/17/commits/6d927acce9329a317e6ce0453a1d5a9471d149ba)

File: authorizerd.go [hunk 2]
```diff
@@ -467,9 +480,59 @@ func (a *authority) authorize(ctx context.Context, m mode, tok, act, res, query
...
+		rolesSize += int64(len(role)) + int64(unsafe.Sizeof(role))
+	}
+
+	authorizedRolesSize := int64(unsafe.Sizeof(p.AuthorizedRoles()))
+	for _, role := range p.AuthorizedRoles() {
+		authorizedRolesSize += int64(len(role)) + int64(unsafe.Sizeof(role))
+	}
+
+	const int64Size = 8
+	timesSize := int64Size * 2
```
[the remark this answers is not in the document]
Comment (fum1h1to) [reply authorizerd.go:516]: fix: [6d927ac](https://github.com/AthenZ/athenz-authorizer/pull/17/commits/6d927acce9329a317e6ce0453a1d5a9471d149ba)

File: authorizerd.go [hunk 3]
```diff
@@ -467,9 +480,59 @@ func (a *authority) authorize(ctx context.Context, m mode, tok, act, res, query
...
+func (a *authority) GetPrincipalCacheLen() int {
+	return a.cache.Len()
+}
+
+// GetPrincipalCacheSize returns memory usage of cached principals
+func (a *authority) GetPrincipalCacheSize() int64 {
+	return int64(a.cache.Size()) + a.cacheMemoryUsage.Load() + int64(a.cacheMemoryUsageMap.Size())
+}
+
+// cacheExpiredHook is a callback that is invoked when a cache entry expires.
```
[the remark this answers is not in the document]
Comment (fum1h1to) [reply authorizerd.go]: fix: [ca13e96](https://github.com/AthenZ/athenz-authorizer/pull/17/commits/ca13e9642087d1969c8b3a16e8503c19e48b1acc)

**Review by y-myajima — APPROVED** [review]
問題ないと思います

**Pull request merged**